src/main/java/fi/aalto/cs/apluscourses/utils/PluginAutoInstaller.java

```diff
@@ -0,0 +1,151 @@
+   * has been cancelled due to missing consent or an error.
+   */
+  @RequiresEdt
+  public static Boolean ensureDependenciesInstalled(@Nullable Project project,
+                                                    @Nullable Notifier notifier,
+                                                    @NotNull List<PluginDependency> pluginNames,
+                                                    @NotNull PluginInstallerCallback callback) {
+    final var pluginIDs = pluginNames.stream().map(p -> PluginId.getId(p.getId())).collect(Collectors.toList());
+
+    // Downloading and enabling plugins are different operations, so we handle these separately.
+    final var pluginsToDownload = pluginIDs.stream()
+        .filter(PluginAutoInstaller::shouldDownloadPlugin).collect(Collectors.toList());
+    final var pluginsToEnable = pluginIDs.stream()
+        .filter(PluginAutoInstaller::shouldEnablePlugin).collect(Collectors.toList());
+
+    pluginsToDownload.forEach(id -> logger.info("Plugin to download: {}", id.getIdString()));
+    pluginsToEnable.forEach(id -> logger.info("Plugin to enable: {}", id.getIdString()));
+
+    if (pluginsToDownload.isEmpty() && pluginsToEnable.isEmpty()) {
+      return true;
+    }
+
+    // allPluginNames = list of PluginDependencies for all plugins in pluginsToDownload or pluginsToEnable
+    final var missingPluginsStream = Stream.concat(pluginsToDownload.stream(), pluginsToEnable.stream());
+    final var allPluginNames = pluginNames.stream()
+        .filter(p -> missingPluginsStream.anyMatch(id -> id.getIdString().equals(p.getId())))
+        .collect(Collectors.toList());
+
+    final var consentResult = callback.askForInstallationConsent(allPluginNames);
+    if (consentResult == PluginInstallerCallback.ConsentResult.IGNORE_INSTALL) {
+      return true; // pretend that everything went fine
+    } else if (consentResult == PluginInstallerCallback.ConsentResult.REJECTED) {
+      return null; // signal to the caller to abort the operation
+    }
+
+    // Enable all requires plugins that are installed, but disabled.
+    pluginsToEnable.forEach(PluginManagerCore::enablePlugin);
+
+    final var downloadablePluginNodes = pluginsToDownload.stream()
+        .map(PluginNode::new).collect(Collectors.toList());
+
+    return ProgressManager.getInstance().runProcessWithProgressSynchronously(() -> {
+      final var indicator = Objects.requireNonNull(ProgressManager.getInstance().getProgressIndicator());
+      indicator.setIndeterminate(true);
+
+      // The semaphore must be released (i.e. permits must be raised) once per each plugin installed.
+      // Only when the semaphore's permit count reaches 1, this function can return.
+      final Semaphore lock = new Semaphore(-downloadablePluginNodes.size() + 1);
+
+      // "true" means that installation is successful (so far)
+      AtomicBoolean installationResult = new AtomicBoolean(true);
+
+      ApplicationManager.getApplication().invokeLater(() -> {
```

Comment:
We invoke this in EDT but the only thing it does is calls a write action, am I right? Do we need EDT to call a write action? :thinking:

Reply:
`runProcessWithProgressSynchronously` must be called from EDT, otherwise the progress bar won't show up as a pop-up window, only as this tiny bar in the right bottom corner of the IDE, in the notification box.
The lambda itself, however, doesn't execute in EDT, so we need the invokeLater inside to run a write action.

Reply:
(note: this comes from my own observations, I didn't find any documentation regarding this behaviour)